### PR TITLE
🌊 - Kim

### PR DESCRIPTION
**Pull request by kimvitug05**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | A custom method I wrote was `get_top_ten` which takes a string `category` and returns a list of up to 10 `Work`s in that category with the highest votes (if there are ties, it will order alphabetically). 
Describe how you approached testing that model method. What edge cases did you come up with? | I made sure to address positive & negative test cases for relationships, validation and all custom methods. An edge case that I came up with was for `get_top_ten` and `get_spotlight` when there are no `Works`. 
What are `session` and `flash`? What is the difference between them? | `session` and `flash` are both hashes that store user data. `session` data persists over multiple requests, whereas `flash` is stored for only one request.
What was one thing that you gained more clarity on through this assignment? | I gained more clarity on nested routes & custom routes.
What is the Heroku URL of your deployed application? | https://kims-media-ranker.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/models/work.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+               .first
+  end
+
+  def self.get_top_ten(category)
+    return Work.where(category: category)
+               .left_joins(:votes)
+               .group(:id)
+               .order('COUNT(votes.work_id) DESC, works.title')
+               .first(10)
+  end
```
Comment (CheezItMan) [inline app/models/work.rb:18]: This is so beautiful I could cry 😭

File: app/models/vote.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+
+  validates_uniqueness_of :user_id, scope: :work_id, message: 'User has already voted for this work'
```
Comment (CheezItMan) [inline app/models/vote.rb:5]: 👍

File: app/views/homepages/_spotlight.html.erb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<section>
```
Comment (CheezItMan) [inline app/views/homepages/_spotlight.html.erb:1]: Great a partial!

File: app/views/homepages/_top10.html.erb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<section>
```
Comment (CheezItMan) [inline app/views/homepages/_top10.html.erb:1]: Another partial!

File: app/views/homepages/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<%= render partial: 'spotlight', locals: { :work => Work.get_spotlight } %>
+
+<section class="top10-container">
+  <%= render partial: 'top10', locals: { :works => Work.get_top_ten('movie'), :category => 'Movies' } %>
+  <%= render partial: 'top10', locals: { :works => Work.get_top_ten('book'), :category => 'Books' } %>
+  <%= render partial: 'top10', locals: { :works => Work.get_top_ten('album'), :category => 'Albums' } %>
```
Comment (CheezItMan) [inline app/views/homepages/index.html.erb:6]: 😍

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    <% if @vote %>
+      <%= render partial: 'shared/flash', :locals => { :errors => @vote.errors } %>
+    <% elsif @user %>
+      <%= render partial: 'shared/flash', :locals => { :errors => @user.errors } %>
+    <% elsif @work %>
+      <%= render partial: 'shared/flash', :locals => { :errors => @work.errors } %>
+    <% else %>
+      <%= render partial: 'shared/flash' %>
+    <% end %>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:23]: Interesting, it works.

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class ApplicationController < ActionController::Base
+  before_action :current_user
+
+  add_flash_types :warning, :success
+
+  def current_user
+    @current_user = User.find_by(id: session[:user_id])
+  end
```
Comment (CheezItMan) [inline app/controllers/application_controller.rb:8]: 👍

File: test/models/work_test.rb [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+      top_movies.each do |work|
+        expect(work).must_be_instance_of Work
+        expect(work[:category]).must_equal "movie"
+      end
+
+      expect(top_movies.first.title).must_equal works(:ace).title
+      expect(top_movies.last.title).must_equal works(:us).title
+    end
+
+    it "if category has more than 10 works, method only returns the first 10" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:90]: Any logic for ties in num votes for top ten?

**Review by CheezItMan — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️, although no seeds in Heroku
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️, great model methods!
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | 💚 Niiiceee






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

## Summary

This is an outstanding submission, you hit all the learning goals here and have a great submission.  I don't really have any areas to suggest for improvement.
